README.md
# minimalistic_bottom_navigation

Minimalistic bottom navigation bar for Flutter projects. Based on my [Dribbble design](https://dribbble.com/shots/6958560-Minimalistic-Bottom-Navigation).

I created this design because I find the standard bottom navigation too boring. There is a lot of cool designs available to use, but most of them relies on icons instead of text. An icon without explanatory text can be confusing to some users, which is why I tried to come up with a design that combines beauty and user experience.

For the best result when using the Minimalistic Bottom Navigation, please use the following style guide:

- The text of the items should be roughly the same length. This is necessary because each item has the same amount of horizontal space on the screen. Fewer items can take bigger differences in length.
- Icons should be solid and have a similar color to the background. This makes the icon visible but not taking too much attention. The opacity property can be used to unite the icon color with the background.

## Demo

![Demo](https://github.com/Friiiis/minimalistic_bottom_navigation/blob/master/example.gif "Demo Gif")

## Installation

In the `dependencies:` section of your `pubspec.yaml`, add the following line:

```yaml
  minimalistic_bottom_navigation: <latest_version>
```

## Usage

```dart
import 'package:minimalistic_bottom_navigation/minimalistic_bottom_navigation.dart';

class _MyHomePageState extends State<MyHomePage> {
  int currentIndex = 0;
  List<Color> colors = List<Color>();

  @override
  void initState() {
    super.initState();
    colors.add(Colors.green);
    colors.add(Colors.redAccent);
    colors.add(Colors.orangeAccent);
  }

  void bottomNavigationTapped(int index) {
    setState(() {
      currentIndex = index;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: AnimatedContainer(
          duration: Duration(milliseconds: 500),
          curve: Curves.ease,
          color: colors[currentIndex],
        ),
      ),
      bottomNavigationBar: MinimalisticBottomBar(
        iconOpacity: 0.2,
        elevation: 0,
        currentIndex: currentIndex,
        onIndexChanged: (index) {
          bottomNavigationTapped(index);
        },
        items: <MinimalisticBottomBarItem>[
          MinimalisticBottomBarItem(
            text: 'Home',
            icon: Icons.home,
            color: colors[0],
          ),
          MinimalisticBottomBarItem(
            text: 'Music',
            icon: Icons.headset,
            color: colors[1],
          ),
          MinimalisticBottomBarItem(
            text: 'Settings',
            icon: Icons.settings,
            color: colors[2],
          ),
        ],
      ),
    );
  }
}

```

## Properties

- **height** `double`: Height of the navigation bar. Defaults to `80`.
- **backgroundColor** `Color`: The background color of the navigation bar. Defaults to `Colors.white`.
- **foregroundColor** `Color`: Color of the font. Defaults to `Color(0xff4a4a4a)`.
- **iconOpacity** `double`: The opacity of the color of the icons, should be between `0` and `1`. Defaults to `0.5`.
- **elevation** `double`: The elevation of the navigation bar. Defaults to `2`.
- **items** `List<MinimalisticBottomBarItem>`: Items in the navigation bar. Must contain between 2 and 5 items (both inclusive).
- **currentIndex** `int`: The current index of the navigation bar.
- **onIndexChanged** `ValueChanged<int>`: Callback when an item is selected.

## Example

View the Flutter app in the `example` directory to see a working example of how to use and implement the slider.

## Contributors

Gustav Friis-Hansen: https://github.com/Friiiis

Stefan Petrovic: https://github.com/StefPetro

CHANGELOG.md
# Changelog

## [0.0.2] - 24/09 - 2019

* Added working example.

## [0.0.1] - 24/09 - 2019

* Initial release.


test/minimalistic_bottom_navigation_test.dart
/*import 'package:flutter_test/flutter_test.dart';

import 'package:minimalistic_bottom_navigation/minimalistic_bottom_navigation.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}*/


example/example/lib/minimalistic_bottom_navigation/src/minimalistic_bottom_bar.dart
import 'package:flutter/material.dart';
import 'minimalistic_bottom_bar_item.dart';

class MinimalisticBottomBar extends StatefulWidget {
  /// Height of the navigation bar. Defaults to 80.
  final double height;

  /// The background color of the navigation bar. Defaults to Colors.white.
  final Color backgroundColor;

  /// Color of the font. Defaults to Color(0xff4a4a4a).
  final Color fontColor;

  /// The opacity of the color of the icons, should be between 0 and 1. Defaults to 0.5.
  final double iconOpacity;

  /// The elevation of the navigation bar. Defaults to 2.
  final double elevation;

  /// Items in the navigation bar. Must be between 2 and 5 (both inclusive).
  final List<MinimalisticBottomBarItem> items;

  /// The current index of the navigation bar.
  final int currentIndex;

  /// Callback when an item is selected.
  final ValueChanged<int> onIndexChanged;

  static const fontHex = Color(0xff4a4a4a);

  MinimalisticBottomBar({
    this.height = 80,
    this.backgroundColor = Colors.white,
    this.fontColor = fontHex,
    this.iconOpacity = 0.5,
    this.elevation = 2,
    @required this.items,
    @required this.currentIndex,
    @required this.onIndexChanged,
  }) : assert(items.length >= 2 &&
            items.length <= 5 &&
            iconOpacity >= 0 &&
            iconOpacity <= 1);

  @override
  _MinimalisticBottomBarState createState() => _MinimalisticBottomBarState();
}

class _MinimalisticBottomBarState extends State<MinimalisticBottomBar>
    with TickerProviderStateMixin {
  bool showingFirstIcon = true;

  Icon firstIcon;
  Icon secondIcon;

  @override
  void initState() {
    super.initState();
    firstIcon = Icon(
      widget.items[widget.currentIndex].icon,
      color: widget.items[widget.currentIndex].color
          .withOpacity(widget.iconOpacity),
      size: widget.height * 0.65,
    );
    secondIcon = Icon(
      widget.items[widget.currentIndex].icon,
      color: widget.items[widget.currentIndex].color
          .withOpacity(widget.iconOpacity),
      size: widget.height * 0.65,
    );
  }

  @override
  Widget build(BuildContext context) {
    double screenWidth = MediaQuery.of(context).size.width;

    if (showingFirstIcon) {
      secondIcon = Icon(
        widget.items[widget.currentIndex].icon,
        color: widget.items[widget.currentIndex].color
            .withOpacity(widget.iconOpacity),
        size: widget.height * 0.6,
      );
      showingFirstIcon = false;
    } else {
      firstIcon = Icon(
        widget.items[widget.currentIndex].icon,
        color: widget.items[widget.currentIndex].color
            .withOpacity(widget.iconOpacity),
        size: widget.height * 0.6,
      );
      showingFirstIcon = true;
    }

    return Container(
      height: widget.height,
      decoration: BoxDecoration(
        color: widget.backgroundColor,
        boxShadow: [
          BoxShadow(
            color: Colors.black26,
            blurRadius: widget.elevation,
            offset: Offset(0, 0),
            spreadRadius: widget.elevation,
          ),
        ],
      ),
      child: Center(
        child: Stack(
          children: <Widget>[
            AnimatedPositioned(
              duration: Duration(milliseconds: 500),
              curve: Curves.fastOutSlowIn,
              left: screenWidth / widget.items.length * widget.currentIndex,
              child: Container(
                width: screenWidth / widget.items.length,
                height: widget.height,
                child: Center(
                  child: AnimatedCrossFade(
                    duration: Duration(milliseconds: 250),
                    firstChild: firstIcon,
                    secondChild: secondIcon,
                    crossFadeState: showingFirstIcon
                        ? CrossFadeState.showFirst
                        : CrossFadeState.showSecond,
                  ),
                ),
              ),
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: widget.items.map((item) {
                int index = widget.items.indexOf(item);
                return InkWell(
                  onTap: () {
                    widget.onIndexChanged(index);
                  },
                  child: Container(
                    width: screenWidth / widget.items.length,
                    child: Center(
                      child: Text(
                        item.text,
                        textAlign: TextAlign.center,
                        style: TextStyle(
                          color: widget.fontColor,
                          fontSize: 16,
                          fontWeight: widget.currentIndex == index
                              ? FontWeight.bold
                              : FontWeight.normal,
                        ),
                      ),
                    ),
                  ),
                );
              }).toList(),
            ),
          ],
        ),
      ),
    );
  }
}


example/example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:minimalistic_bottom_navigation/minimalistic_bottom_navigation.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int currentIndex = 0;
  List<Color> colors = List<Color>();

  @override
  void initState() {
    super.initState();
    colors.add(Colors.green);
    colors.add(Colors.redAccent);
    colors.add(Colors.blueAccent);
  }

  void bottomNavigationTapped(int index) {
    setState(() {
      currentIndex = index;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: AnimatedContainer(
          duration: Duration(milliseconds: 500),
          curve: Curves.ease,
          color: colors[currentIndex],
        ),
      ),
      bottomNavigationBar: MinimalisticBottomBar(
        iconOpacity: 0.2,
        elevation: 0,
        currentIndex: currentIndex,
        onIndexChanged: (index) {
          bottomNavigationTapped(index);
        },
        items: <MinimalisticBottomBarItem>[
          MinimalisticBottomBarItem(
            text: 'Home',
            icon: Icons.home,
            color: colors[0],
          ),
          MinimalisticBottomBarItem(
            text: 'Music',
            icon: Icons.headset,
            color: colors[1],
          ),
          MinimalisticBottomBarItem(
            text: 'Settings',
            icon: Icons.settings,
            color: colors[2],
          ),
        ],
      ),
    );
  }
}
